File: clojure/queue.clj
(ns queue )

(defprotocol ImplementQueue
             (enqueue [this item])
             (dequeue [this])
             (is-empty? [this])
             (top [this])
             )

(defrecord Queue [elements]
           ImplementQueue
           (enqueue
             [_ item]
             (swap! elements conj item))
           (dequeue
             [_]
             (swap! elements drop-last))
           (is-empty?
             [_]
             (= 0 (count @elements)))
           (top
             [_]
             (first @elements)))

(def queue (Queue. (atom '())))


(println(.is-empty? queue))
(println(.enqueue queue 5))
(println(.enqueue queue 1))
(println(.enqueue queue 6))
(println(.enqueue queue 9))
(println(.is-empty? queue))
(println(.dequeue queue))
(println queue)
(println(.top queue))

File: java/Queue.java
import Exception.FullQueue;
import Exception.EmptyQueue;

public class Queue<E> {
    private int size;
    private int top;
    private int first;
    public E[] elements;

    public Queue(int s){
        first = 0;
        size = s;
        top = -1;

        if (size < 0) {
            throw new IllegalArgumentException();
        }

        elements = (E[]) new Object[size];
    }

    public boolean isEmpty(){
        return top == -1;
    }

    public boolean isFull(){
        return top == size - 1;
    }

    public void enqueue(E element){
        if (top == size - 1) {
            throw new FullQueue();
        }
        elements[++top] = element;
    }

    public void dequeue(){
        if (top == -1) {
            throw new EmptyQueue();
        }
        elements[first] = elements[first + 1];
        elements[first + 1] = null;
    }
}
